### DISCUSSION TITLE
Which builder to use with a prebuilt golang binary if I still want Dokku to manage the deployment?

### DISCUSSION THREAD

**Discussion by bakatz** [dokku/dokku#7056] [Q&A] [answered]
Hi there, to give some context to that question: I have a prebuilt, statically linked golang binary for my app. I use GitHub Actions for the build and want Dokku to handle deployment only. I don't want to use system resources having Dokku doing builds on my VPS as I already get free build minutes with GH Actions if that makes sense.

In this situation, is this the recommended approach?
1. Builder: https://dokku.com/docs/deployment/builders/null (I'm not sure about this one because it says "should not be used in normal operation" - I want Dokku to treat this as any other app to be able to use the logs, scale commands and such)
2. Deployment: https://dokku.com/docs/deployment/methods/archive (I don't really need a Docker image deployment since everything is statically linked)

Thanks!

**Comment by josegonzalez** [comment]
Dokku requires using containers for the artifact we deploy, no way around that. You will probably need to wrap your compiled binary in a docker image - maybe `FROM scratch` as your base if everything is statically compiled - and deploy the resulting image.

The null builder is usually there if you want to have Dokku link out to something running on the host - like Dokku Pro - though Dokku will not handle the lifecycle of whatever that thing is.

**Reply by bakatz** [reply]
> Dokku requires using containers for the artifact we deploy, no way around that. You will probably need to wrap your compiled binary in a docker image - maybe `FROM scratch` as your base if everything is statically compiled - and deploy the resulting image.
> 
> The null builder is usually there if you want to have Dokku link out to something running on the host - like Dokku Pro - though Dokku will not handle the lifecycle of whatever that thing is.

Got it @josegonzalez, thanks for the response. It's not much of a problem to wrap the binary in a "distroless" or "scratch" docker image like you mentioned, I can do that.

In that case though, which "builder" should be used? I don't need Dokku to re-build the image again, I just need it to pull and run the image from the registry - and hopefully manage it as it does any other app.

It seems like this command would be the right approach? `dokku git:from-image my-app my-registry/my-app:latest` from https://dokku.com/docs/deployment/methods/image/#initializing-an-app-repository-from-a-docker-image

Will that automatically set up Dokku to just run the existing Docker image instead of trying to build? And does that need to be run on every CI run to refresh the latest image?

**Reply by josegonzalez** [reply, marked as answer]
You don't need a builder, those are just for if you push code via git. You can use the `git:from-image` command, though I recommend using a specific tag for each deploy or the image digest rather than just `latest` as you'll find issues doing that.

And yes, you need to run that from CI to trigger a deploy, Dokku won't run a binary to listen for external updates (IIRC there isn't really an API for this anyways on the Docker Registry spec). We have a number of CI integrations that may help set this up for you.

**Reply by bakatz** [reply]
Perfect, thank you. Just found the GH actions step that you all maintain - will be using that, looks simple. Giving Dokku a shot tomorrow and will likely be writing a blog post with how it goes 👍
